mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/work/impl/SearchIndexingPlanFilterContextImpl.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.pojo.work.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.hibernate.search.mapper.pojo.logging.impl.Log;
+import org.hibernate.search.mapper.pojo.model.spi.PojoRawTypeIdentifier;
+import org.hibernate.search.mapper.pojo.scope.impl.PojoScopeTypeContextProvider;
+import org.hibernate.search.mapper.pojo.work.SearchIndexingPlanFilterContext;
+import org.hibernate.search.mapper.pojo.work.spi.ConfiguredSearchIndexingPlanFilter;
+import org.hibernate.search.util.common.data.spi.KeyValueProvider;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+public class SearchIndexingPlanFilterContextImpl implements SearchIndexingPlanFilterContext {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final PojoScopeTypeContextProvider contextProvider;
+
+	private final KeyValueProvider<Class<?>, PojoRawTypeIdentifier<?>> typeIdentifierByClass;
+
+	private final Set<PojoRawTypeIdentifier<?>> includes = new HashSet<>();
+	private final Set<PojoRawTypeIdentifier<?>> excludes = new HashSet<>();
+
+	public SearchIndexingPlanFilterContextImpl(PojoScopeTypeContextProvider typeManager) {
+		this.contextProvider = typeManager;
+
+		Map<Class<?>, PojoRawTypeIdentifier<?>> typeIdentifierByClassContent = new LinkedHashMap<>();
+		for ( PojoRawTypeIdentifier<?> identifier : typeManager.allNonInterfaceSuperTypes() ) {
+			Class<?> klass = identifier.javaClass();
+			// at this point we only care about classes since we'll be using this map to find identifiers by a class
+			// in include(Class)/exclude(Class) methods.
+			// for any dynamic types we are expecting users to use the include(String)/exclude(String) instead.
+			if ( !klass.isInterface() ) {
```

Comment:
mmmm right.... me coming up with confusing behaviour and naming again 🙈 😄 sorry 
that should've been something like - `all-supertypes-of-indexed-and-contained-types-excluding-interfaces-unless-those-interfaces-are-named-interfaces-coming-from-dynamic-type-definitions`, but that name might be a bit too long even for Java 🙈

Overall the idea is that when a user calls include/exclude -- we want to convert a string/class to a type identifier. We cannot get the entire hierarchy at that point since if we do so and store it in the include/exclude collections, we can end up throwing an exception when the user starts to add some other sub/super types to the filter as those would already be in the collections. 
So basically, we need a converter from string/class to type identifier. And this converter should allow for `Object` and all other supertypes we want to be included.
Then once we are done collecting info from the user, we need a type-identifier -> subtypes so that we can build an actual filter. 

And here, this `KeyValueProvider` is only for those class methods, so I didn't want it to have a `Map` in there too... Alternatively, we could say that we would allow Map here and that if the user passes it then we include/exclude all those dynamic types... but I thought that what if a user decides to create their own entity and implement a map on it ... so I didn't want to go that route 😃

Reply:
I'm having trouble understanding.

Let's try it another way: if you remove this check, what breaks?

Reply:
If we'd remove this, then <Map.class, "OneOfDynamicTypes"> will be in the `KeyValueProvider`.
no current tests will fail. But we'd be able to make a call like:
```
context.exclude(Map.class)
```
and it will not fail. But since this `KeyValueProvider` only would point to one of the dynamic types we won't be able to say which type exactly will be excluded ....

Reply:
Ok I see. Maybe you need such a test?

Anyway, why don't you restore the  `KeyValueProvider<Class, Set<Type context>>` that you had in previous iterations, and make sure to only put non-named types in there ? After all, there is no reason to build this map every single time you instantiate a context, since you can build it on startup and make it a key value provider.

Reply:
+1 added a test. 
As for the `KeyValueProvider<Class, Set<Type context>> ` ... 
I could do something like:
```java
private final Map<Class<?>, PojoRawTypeModel<?>> indexedContainedSuperType = new LinkedHashMap<>();

public <E> void addIndexed(PojoRawTypeModel<E> typeModel, PojoIndexedTypeManager<?, E> typeManager) {
	indexed.add( typeManager );
        // .....
	typeModel.descendingSuperTypes()
			.filter( type -> !type.typeIdentifier().javaClass().isInterface() )
			.forEach( type -> {
				indexedContainedSuperType.put( type.typeIdentifier().javaClass(), type );
			} );
}
```
But that wouldn't be a `AbstractPojoTypeManager` or other similar `Type context`s. Or I can do a similar to what we have here - `KeyValueProvider<Class, Set<TypeIdentifier>>`, but then as you've mentioned earlier it wouldn't really be a `typeManagerContainer`.
This `KeyValueProvider` must contain a pair `<Object.class, "anything that we can get a TypeIdentifier<Object> from">`, since its only purpose is to convert `Class<?>` passed into `SearchIndexingPlanFilterContext#inlcude(Class)`/`SearchIndexingPlanFilterContext#exclude(Class)` to a `PojoRawTypeIdentifier`. We never have a `PojoIndexedTypeManager` for the `Object.class` right?

Reply:
Ok, I just had a closer look at why you even need this map and... you can just use `PojoRawTypeIdentifier.of(clazz)` xD

It's not great, but that's what we do elsewhere already (`org.hibernate.search.mapper.pojo.standalone.mapping.impl.StandalonePojoMapping#createScope`).

As to failing when the class is an interface... I think it all boils down to failing when the type identifier doesn't match anything in `contextProvider.allNonInterfaceSuperTypes()`. And if there's a named type using an interface, its type id will never match `PojoRawTypeIdentifier.of(clazz)` (since that results in an unnamed type ID).

By the way I also noticed that `org.hibernate.search.mapper.pojo.scope.impl.PojoScopeTypeContextProvider#allByNonInterfaceSuperType` is implemented to rely on a `Map` (`allByNonInterfaceSuperType`) and return an empty optional when there's no match, but I don't think this should ever happen? We definitely expect everything passed to `include`/`exclude` to have an effect.

To sum up I'd suggest the following:

* don't build a map every single time you create a `SearchIndexingPlanFilterContextImpl`
* use `PojoRawTypeIdentifier.of(clazz)` to turn a class into a type id
* add a check to `include(Class)`/`exclude(Class)` that the type identifier is present in `allNonInterfaceSuperTypes()`. If it's not, you can throw `log.unknownClassForNonInterfaceSuperType()`. If necessary introduce a `allNonInterfaceSuperClasses()` that returns a set of `Class` instead of type identifiers, but that's mostly for the error message.
* add a test with `.include(Integer.class)`/`.exclude(...)` and check that it fails with a meaningful and appropriate message.
* [probably already done] add a test with `.include(someinterface)`/`.exclude(...)` and check that it fails with a meaningful and appropriate message
* [probably already done] add a test with `.include("someNameThatDoesNotMatchAnEntity")`/`.exclude(...)` and check that it fails with a meaningful and appropriate message

Also, unrelated, but I just thought about it:

* Add a test with `.include("validEntityName")`/`.exclude(...)` but with the entity name pointing to an entity that has absolutely no Hibernate Search annotations and is not `@IndexedEmbedded` anywhere either... but it has indexed subclasses. This is expected to work, I think? But I don't think it will.

Does all that make sense?

Reply:
> you can just use PojoRawTypeIdentifier.of(clazz) xD

heyy !! that's a cheatcode 😆 ahah
I've seen that method before, and I thought that it wouldn't work if we had an indexed entity with a name specified, (I was thinking that in that case, the name won't be null and hence equals wouldn't work, and we wouldn't find anything in the keyvalue providers ...) ohhhhh I should've just tried that to make sure that it wasn't actually so 🙈 🤦🏻‍♂️ 😆 
and yes all makes sense, thanks 😃

Reply:
> Add a test with .include("validEntityName")/.exclude(...) but with the entity name pointing to an entity that has absolutely no Hibernate Search annotations and is not @IndexedEmbedded anywhere either... but it has indexed subclasses. This is expected to work, I think? But I don't think it will.

Right, so this didn't work as you've been suspecting 😃 🎉.

There are two commits with new changes. The first one addresses your suggestions and adds a new key-value provider for the entity name since existing ones do not cover that last test case... I did put it in `PojoTypeManagerContainer` even though it is storing the `PojoRawTypeModel`, but I guess we won't have anything else for entity-supertypes", right?
The last one adds more tests for dynamic type cases when we try to pass names of a supertype of an indexed subtype or just a name of an entity that is not mapped in HSEARCH.